javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -9,38 +9,53 @@
 import javaslang.collection.VectorModule.Combinations;
 import javaslang.control.Option;
 
+import java.io.Serializable;
 import java.util.*;
 import java.util.Set;
 import java.util.function.*;
 import java.util.stream.Collector;
 
+import static javaslang.collection.Arrays2.*;
 import static javaslang.collection.Collections.*;
+import static javaslang.collection.Vector.Helper.fullTreeSize;
+import static javaslang.collection.Vector.VectorTree.emptyTree;
 
 /**
- * Vector is the default Seq implementation. It provides the best performance in between Array (with constant time element access)
- * and List (with constant time element addition).
+ * Vector is the default Seq implementation that provides effectively constant time access to any element.
+ * Many other operations (e.g. `tail`, `drop`, `slice`) are also effectively constant.
+ *
+ * It's implemented using a `bit-mapped vector trie`, i.e. a tree where each node has 32 children.
+ * For integers this means that the tree's depth will be ≤ 6.
+ *
+ * For symmetrically fast beginning and end operations (e.g. `drop` vs `dropRight`) two small arrays are used as depthless staging areas.
+ * The first and last arrays have dynamic lengths (are shifted to the correct index on drop), but the middle one has internal offset to avoid shifting.
+ * The first one is only empty if everything is empty; if the last is empty, so is the middle.
  *
  * @param <T> Component type of the Vector.
- * @author Ruslan Sennov
- * @since 2.0.0
+ * @author Pap Lőrinc
+ * @since 3.0.0
  */
-
 @SuppressWarnings({"SuspiciousArrayCast", "unchecked"})
 public class Vector<T> implements Kind1<Vector<?>, T>, IndexedSeq<T> {
     private static final long serialVersionUID = 1L;
 
-    private static final Vector<?> EMPTY = new Vector<>(HashArrayMappedTrie.empty());
+    static int BRANCHING_BASE = 5;
```

Comment:
it's changeable from code, is used from tests to validate multiple levels.
In Scala they're not testing the upper levels, as it would take too much time
(it wasn't even functional in Scala for 6 years, when someone finally gave it a try for billions of elements)
```scala
display4 = display5((newIndex >> 20) & 31).asInstanceOf[Array[AnyRef]] // 20 -> should be 25
if (display4 == null) display4 = new Array(32)
display3 = display4((newIndex >> 20) & 31).asInstanceOf[Array[AnyRef]]
if (display3 == null) display3 = new Array(32)
display2 = display3((newIndex >> 15) & 31).asInstanceOf[Array[AnyRef]]
```